ambry-store/src/main/java/com.github.ambry.store/LogSegment.java

```diff
@@ -0,0 +1,339 @@
+/**
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.store;
+
+import com.github.ambry.utils.Pair;
+import com.github.ambry.utils.Utils;
+import java.io.File;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.channels.ReadableByteChannel;
+import java.util.concurrent.atomic.AtomicLong;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Represents a segment of a log. The segment is represented by its relative position in the log and the generation
+ * number of the segment. Each segment knows the segment that "follows" it logically (if such a segment exists) and can
+ * transparently redirect operations if required. A segment can be in one of the states represented by {@link State}.
+ */
+class LogSegment implements Read, Write {
+  /**
+   * Used to describe the state of the LogSegment.
+   */
+  enum State {
+    /**
+     * The LogSegment is completely free and no writes have been accommodated in the segment.
+     */
+    FREE,
+    /**
+     * The LogSegment is currently being written to. It still has capacity remaining for more writes.
+     */
+    ACTIVE,
+    /**
+     * The LogSegment is full and is accepting no more writes.
+     */
+    SEALED
+  }
+
+  private final FileChannel fileChannel;
+  private final File file;
+  private final long capacityInBytes;
+  private final String name;
+  private final Pair<File, FileChannel> segmentView;
+  private final StoreMetrics metrics;
+  private final AtomicLong endOffset;
+  private final AtomicLong refCount = new AtomicLong(0);
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+
+  /**
+   * The state of the segment.
+   * <p/>
+   * The natural state transition of a LogSegment is FREE -> ACTIVE -> SEALED but it can be externally manipulated to
+   * move in any direction.
+   */
+  State state;
+  /**
+   * Reference to the segment that logically "follows" this segment.
+   */
+  LogSegment next = null;
```

Comment:
no, this isn't being used right now. The log will need to manipulate this and I didn't see any reason for adding getter/setter because they wouldn't do anything special. I don't see a reason to not allow components to set it directly.